File: testing/unit-tests/pkmntest/cpp/pokemon_subclass.cpp
/*
 * Copyright (c) 2018 Nicholas Corgan (n.corgan@gmail.com)
 *
 * Distributed under the MIT License (MIT) (See accompanying file LICENSE.txt
 * or copy at http://opensource.org/licenses/MIT)
 */

#include <pkmntest/pokemon_subclass.hpp>

#include <pkmn/enums/enum_to_string.hpp>

namespace pkmntest
{
    pokemon_subclass::pokemon_subclass(
        pkmn::e_species species,
        pkmn::e_game game
    ): pkmn::pokemon(),
       _database_entry(species, game, ""),
       _species(species),
       _game(game),
       _condition(pkmn::e_condition::NONE),
       _gender(pkmn::e_gender::MALE),
       _held_item(pkmn::e_item::NONE),
       _nature(pkmn::e_nature::HARDY),
       _original_trainer_gender(pkmn::e_gender::MALE),
       _language(pkmn::e_language::ENGLISH),
       _ability(pkmn::e_ability::NONE),
       _ball(pkmn::e_ball::POKE_BALL),
       _original_game(pkmn::e_game::NONE),
       _is_egg(false),
       _is_shiny(false),
       _pokerus_duration(0),
       _current_trainer_friendship(0),
       _level_met(0),
       _personality(0),
       _experience(0),
       _level(5),
       _current_hp(0),
       _nickname(pkmn::species_to_string(species)),
       _original_trainer_name("Test"),
       _location_met(),
       _location_met_as_egg(),
       _original_trainer_id({0}),
       _markings(),
       _ribbons(),
       _contest_stats(),
       _EVs(),
       _IVs(),
       _stats(),
       _moves()
    {
        // This calls an implementation-agnostic function that populates
        // values with suitable defaults to match in-game policy.
        initialize_default_values(species, game);
    }

    pkmn::pokemon::sptr pokemon_subclass::clone() const
    {
        return std::make_shared<pokemon_subclass>(*this);
    }

    pkmn::pokemon::sptr pokemon_subclass::to_game(pkmn::e_game game)
    {
        return std::make_shared<pokemon_subclass>(_species, game);
    }

    void pokemon_subclass::export_to_file(const std::string& filepath)
    {
        (void)filepath;
    }

    pkmn::e_species pokemon_subclass::get_species() const
    {
        return _species;
    }

    pkmn::e_game pokemon_subclass::get_game() const
    {
        return _game;
    }

    std::string pokemon_subclass::get_form() const
    {
        return _database_entry.get_form();
    }

    void pokemon_subclass::set_form(const std::string& form)
    {
        _database_entry.set_form(form);
    }

    bool pokemon_subclass::is_egg() const
    {
        return _is_egg;
    }

    void pokemon_subclass::set_is_egg(bool is_egg)
    {
        _is_egg = is_egg;
    }

    const pkmn::database::pokemon_entry& pokemon_subclass::get_database_entry() const
    {
        return _database_entry;
    }

    pkmn::e_condition pokemon_subclass::get_condition() const
    {
        return _condition;
    }

    void pokemon_subclass::set_condition(pkmn::e_condition condition)
    {
        _condition = condition;
    }

    std::string pokemon_subclass::get_nickname() const
    {
        return _nickname;
    }

    void pokemon_subclass::set_nickname(const std::string& nickname)
    {
        _nickname = nickname;
    }

    pkmn::e_gender pokemon_subclass::get_gender() const
    {
        return _gender;
    }

    void pokemon_subclass::set_gender(pkmn::e_gender gender)
    {
        _gender = gender;
    }

    bool pokemon_subclass::is_shiny() const
    {
        return _is_shiny;
    }

    void pokemon_subclass::set_shininess(bool is_shiny)
    {
        _is_shiny = is_shiny;
    }

    pkmn::e_item pokemon_subclass::get_held_item() const
    {
        return _held_item;
    }

    void pokemon_subclass::set_held_item(pkmn::e_item held_item)
    {
        _held_item = held_item;
    }

    pkmn::e_nature pokemon_subclass::get_nature() const
    {
        return _nature;
    }

    void pokemon_subclass::set_nature(pkmn::e_nature nature)
    {
        _nature = nature;
    }

    int pokemon_subclass::get_pokerus_duration() const
    {
        return _pokerus_duration;
    }

    void pokemon_subclass::set_pokerus_duration(int pokerus_duration)
    {
        _pokerus_duration = pokerus_duration;
    }

    std::string pokemon_subclass::get_original_trainer_name() const
    {
        return _original_trainer_name;
    }

    void pokemon_subclass::set_original_trainer_name(const std::string& original_trainer_name)
    {
        _original_trainer_name = original_trainer_name;
    }

    uint16_t pokemon_subclass::get_original_trainer_public_id() const
    {
        return _original_trainer_id.public_id;
    }

    uint16_t pokemon_subclass::get_original_trainer_secret_id() const
    {
        return _original_trainer_id.secret_id;
    }

    uint32_t pokemon_subclass::get_original_trainer_id() const
    {
        return _original_trainer_id.id;
    }

    void pokemon_subclass::set_original_trainer_public_id(
        uint16_t original_trainer_public_id
    )
    {
        _original_trainer_id.public_id = original_trainer_public_id;
    }

    void pokemon_subclass::set_original_trainer_secret_id(
        uint16_t original_trainer_secret_id
    )
    {
        _original_trainer_id.secret_id = original_trainer_secret_id;
    }

    void pokemon_subclass::set_original_trainer_id(
        uint32_t original_trainer_id
    )
    {
        _original_trainer_id.id = original_trainer_id;
    }

    pkmn::e_gender pokemon_subclass::get_original_trainer_gender() const
    {
        return _original_trainer_gender;
    }

    void pokemon_subclass::set_original_trainer_gender(
        pkmn::e_gender original_trainer_gender
    )
    {
        _original_trainer_gender = original_trainer_gender;
    }

    pkmn::e_language pokemon_subclass::get_language() const
    {
        return _language;
    }

    void pokemon_subclass::set_language(pkmn::e_language language)
    {
        _language = language;
    }

    int pokemon_subclass::get_current_trainer_friendship() const
    {
        return _current_trainer_friendship;
    }

    void pokemon_subclass::set_current_trainer_friendship(int current_trainer_friendship)
    {
        _current_trainer_friendship = current_trainer_friendship;
    }

    pkmn::e_ability pokemon_subclass::get_ability() const
    {
        return _ability;
    }

    void pokemon_subclass::set_ability(pkmn::e_ability ability)
    {
        _ability = ability;
    }

    pkmn::e_ball pokemon_subclass::get_ball() const
    {
        return _ball;
    }

    void pokemon_subclass::set_ball(pkmn::e_ball ball)
    {
        _ball = ball;
    }

    int pokemon_subclass::get_level_met() const
    {
        return _level_met;
    }

    void pokemon_subclass::set_level_met(int level_met)
    {
        _level_met = level_met;
    }

    std::string pokemon_subclass::get_location_met(bool as_egg) const
    {
        return as_egg ? _location_met_as_egg : _location_met;
    }

    void pokemon_subclass::set_location_met(
        const std::string& location,
        bool as_egg
    )
    {
        if(as_egg)
        {
            _location_met_as_egg = location;
        }
        else
        {
            _location_met = location;
        }
    }

    pkmn::e_game pokemon_subclass::get_original_game() const
    {
        return _original_game;
    }

    void pokemon_subclass::set_original_game(pkmn::e_game original_game)
    {
        _original_game = original_game;
    }

    uint32_t pokemon_subclass::get_personality() const
    {
        return _personality;
    }

    void pokemon_subclass::set_personality(uint32_t personality)
    {
        _personality = personality;
    }

    int pokemon_subclass::get_experience() const
    {
        return _experience;
    }

    void pokemon_subclass::set_experience(int experience)
    {
        _experience = experience;
    }

    int pokemon_subclass::get_level() const
    {
        return _level;
    }

    void pokemon_subclass::set_level(int level)
    {
        _level = level;
    }

    const std::map<pkmn::e_marking, bool>& pokemon_subclass::get_markings() const
    {
        return _markings;
    }

    void pokemon_subclass::set_marking(
        pkmn::e_marking marking,
        bool value
    )
    {
        _markings[marking] = value;
    }

    const std::map<std::string, bool>& pokemon_subclass::get_ribbons() const
    {
        return _ribbons;
    }

    void pokemon_subclass::set_ribbon(
        const std::string& ribbon,
        bool value
    )
    {
        _ribbons[ribbon] = value;
    }

    const std::map<pkmn::e_contest_stat, int>& pokemon_subclass::get_contest_stats() const
    {
        return _contest_stats;
    }

    void pokemon_subclass::set_contest_stat(
        pkmn::e_contest_stat stat,
        int value
    )
    {
        _contest_stats[stat] = value;
    }

    const pkmn::move_slots_t& pokemon_subclass::get_moves() const
    {
        return _moves;
    }

    void pokemon_subclass::set_move(
        pkmn::e_move move,
        int index
    )
    {
        _moves[index].move = move;
    }

    void pokemon_subclass::set_move_pp(
        int index,
        int pp
    )
    {
        _moves[index].pp = pp;
    }

    const std::map<pkmn::e_stat, int>& pokemon_subclass::get_EVs() const
    {
        return _EVs;
    }

    void pokemon_subclass::set_EV(
        pkmn::e_stat stat,
        int value
    )
    {
        _EVs[stat] = value;
    }

    const std::map<pkmn::e_stat, int>& pokemon_subclass::get_IVs() const
    {
        return _IVs;
    }

    void pokemon_subclass::set_IV(
        pkmn::e_stat stat,
        int value
    )
    {
        _IVs[stat] = value;
    }

    int pokemon_subclass::get_current_hp() const
    {
        return _current_hp;
    }

    void pokemon_subclass::set_current_hp(int current_hp)
    {
        _current_hp = current_hp;
    }

    const std::map<pkmn::e_stat, int>& pokemon_subclass::get_stats() const
    {
        return _stats;
    }

    std::string pokemon_subclass::get_icon_filepath() const
    {
        return "";
    }

    std::string pokemon_subclass::get_sprite_filepath() const
    {
        return "";
    }

    const void* pokemon_subclass::get_native() const
    {
        return nullptr;
    }
}

File: testing/unit-tests/pkmntest/csharp/PokemonSubclass.cs
/*
 * Copyright (c) 2018 Nicholas Corgan (n.corgan@gmail.com)
 *
 * Distributed under the MIT License (MIT) (See accompanying file LICENSE.txt
 * or copy at http://opensource.org/licenses/MIT)
 */

using System;
using NUnit.Framework;

namespace PKMNTest
{

public class PokemonSubclass: PKMN.Pokemon
{
    public PokemonSubclass(
        PKMN.Species species,
        PKMN.Game game
    ): base()
    {
        InitializeDefaultValues(species, game);

        _species = species;
        _game = game;
    }

    override public PKMN.Pokemon Clone()
    {
        return new PokemonSubclass(this.Species, this.Game);
    }

    override public PKMN.Pokemon ToGame(PKMN.Game game)
    {
        return new PokemonSubclass(this.Species, game);
    }

    override public void ExportToFile(string filepath)
    {
    }

    override protected PKMN.Species GetSpecies()
    {
        return _species;
    }

    override protected PKMN.Game GetGame()
    {
        return _game;
    }

    override protected string GetForm()
    {
        return GetDatabaseEntryInternal().Form;
    }

    override protected void SetForm(string form)
    {
        GetDatabaseEntryInternal().Form = form;
    }

    override protected bool GetIsEgg()
    {
        return _isEgg;
    }

    override protected void SetIsEgg(bool isEgg)
    {
        _isEgg = isEgg;
    }

    override protected PKMN.Condition GetCondition()
    {
        return _condition;
    }

    override protected void SetCondition(PKMN.Condition condition)
    {
        _condition = condition;
    }

    override protected string GetNickname()
    {
        return _nickname;
    }

    override protected void SetNickname(string nickname)
    {
        _nickname = nickname;
    }

    override protected PKMN.Gender GetGender()
    {
        return _gender;
    }

    override protected void SetGender(PKMN.Gender gender)
    {
        _gender = gender;
    }

    override protected bool GetIsShiny()
    {
        return _isShiny;
    }

    override protected void SetShininess(bool isShiny)
    {
        _isShiny = isShiny;
    }

    override protected PKMN.Item GetHeldItem()
    {
        return _heldItem;
    }

    override protected void SetHeldItem(PKMN.Item heldItem)
    {
        _heldItem = heldItem;
    }

    override protected PKMN.Nature GetNature()
    {
        return _nature;
    }

    override protected void SetNature(PKMN.Nature nature)
    {
        _nature = nature;
    }

    override protected int GetPokerusDuration()
    {
        return _pokerusDuration;
    }

    override protected void SetPokerusDuration(int pokerusDuration)
    {
        _pokerusDuration = pokerusDuration;
    }

    override protected string GetOriginalTrainerName()
    {
        return _originalTrainerName;
    }

    override protected void SetOriginalTrainerName(string originalTrainerName)
    {
        _originalTrainerName = originalTrainerName;
    }

    override protected uint GetOriginalTrainerID()
    {
        return _originalTrainerID;
    }

    override protected void SetOriginalTrainerID(uint originalTrainerID)
    {
        _originalTrainerID = originalTrainerID;
    }

    override protected ushort GetOriginalTrainerPublicID()
    {
        return (ushort)(_originalTrainerID & 0xFFFF);
    }

    override protected void SetOriginalTrainerPublicID(ushort originalTrainerPublicID)
    {
        _originalTrainerID &= 0xFFFF0000;
        _originalTrainerID |= (uint)originalTrainerPublicID;
    }

    override protected ushort GetOriginalTrainerSecretID()
    {
        return (ushort)(_originalTrainerID >> 16);
    }

    override protected void SetOriginalTrainerSecretID(ushort originalTrainerSecretID)
    {
        _originalTrainerID &= 0xFFFF;
        _originalTrainerID |= ((uint)(originalTrainerSecretID) << 16);
    }

    override protected PKMN.Gender GetOriginalTrainerGender()
    {
        return _originalTrainerGender;
    }

    override protected void SetOriginalTrainerGender(PKMN.Gender originalTrainerGender)
    {
        _originalTrainerGender = originalTrainerGender;
    }

    override protected PKMN.Language GetLanguage()
    {
        return _language;
    }

    override protected void SetLanguage(PKMN.Language language)
    {
        _language = language;
    }

    override protected int GetCurrentTrainerFriendship()
    {
        return _currentTrainerFriendship;
    }

    override protected void SetCurrentTrainerFriendship(int currentTrainerFriendship)
    {
        _currentTrainerFriendship = currentTrainerFriendship;
    }

    override protected PKMN.Ability GetAbility()
    {
        return _ability;
    }

    override protected void SetAbility(PKMN.Ability ability)
    {
        _ability = ability;
    }

    override protected PKMN.Ball GetBall()
    {
        return _ball;
    }

    override protected void SetBall(PKMN.Ball ball)
    {
        _ball = ball;
    }

    override protected int GetLevelMet()
    {
        return _levelMet;
    }

    override protected void SetLevelMet(int levelMet)
    {
        _levelMet = levelMet;
    }

    override protected string GetLocationMet()
    {
        return _locationMet;
    }

    override protected void SetLocationMet(string locationMet)
    {
        _locationMet = locationMet;
    }

    override protected string GetLocationMetAsEgg()
    {
        return _locationMetAsEgg;
    }

    override protected void SetLocationMetAsEgg(string locationMetAsEgg)
    {
        _locationMetAsEgg = locationMetAsEgg;
    }

    override protected PKMN.Game GetOriginalGame()
    {
        return _originalGame;
    }

    override protected void SetOriginalGame(PKMN.Game originalGame)
    {
        _originalGame = originalGame;
    }

    override protected uint GetPersonality()
    {
        return _personality;
    }

    override protected void SetPersonality(uint personality)
    {
        _personality = personality;
    }

    override protected int GetExperience()
    {
        return _experience;
    }

    override protected void SetExperience(int experience)
    {
        _experience = experience;
    }

    override protected int GetLevel()
    {
        return _level;
    }

    override protected void SetLevel(int level)
    {
        _level = level;
    }

    override protected int GetCurrentHP()
    {
        return _currentHP;
    }

    override protected void SetCurrentHP(int currentHP)
    {
        _currentHP = currentHP;
    }

    // Just set the abstraction.
    override protected void SetMarking(PKMN.Marking marking, bool value)
    {
        SetMarkingsMapValueInternal(marking, value);
    }

    // Just set the abstraction.
    override protected void SetRibbon(string ribbon, bool value)
    {
        SetRibbonsMapValueInternal(ribbon, value);
    }

    // Just set the abstraction.
    override protected void SetContestStat(
        PKMN.ContestStat contestStat,
        int value
    )
    {
        SetContestStatsMapValueInternal(contestStat, value);
    }

    // Just set the abstraction.
    override protected void SetMove(
        PKMN.Move move,
        int index
    )
    {
        SetMovesListMoveValueInternal(index, move);
    }

    // Just set the abstraction.
    override protected void SetMovePP(
        int index,
        int PP
    )
    {
        SetMovesListPPValueInternal(index, PP);
    }

    // Just set the abstraction.
    override protected void SetEV(
        PKMN.Stat stat,
        int value
    )
    {
        SetEVsMapValueInternal(stat, value);
    }

    // Just set the abstraction.
    override protected void SetIV(
        PKMN.Stat stat,
        int value
    )
    {
        SetIVsMapValueInternal(stat, value);
    }

    // Internals
    private PKMN.Species _species;
    private PKMN.Game _game;
    private bool _isEgg;
    private PKMN.Condition _condition;
    private string _nickname;
    private PKMN.Gender _gender;
    private bool _isShiny;
    private PKMN.Item _heldItem;
    private PKMN.Nature _nature;
    private int _pokerusDuration;
    private string _originalTrainerName;
    private uint _originalTrainerID;
    private PKMN.Gender _originalTrainerGender;
    private PKMN.Language _language;
    private int _currentTrainerFriendship;
    private PKMN.Ability _ability;
    private PKMN.Ball _ball;
    private int _levelMet;
    private string _locationMet;
    private string _locationMetAsEgg;
    private PKMN.Game _originalGame;
    private uint _personality;
    private int _experience;
    private int _level;
    private int _currentHP;
}

}
